File: Minimize Max Distance to Gas Station.cpp
class Solution {
  public:
    double findSmallestMaxDist(vector<int> &stations, int k) {
        int n = stations.size();
        double left = 0, right = stations[n-1] - stations[0];

        auto canPlaceStations = [&](double maxDist) {
            int neededStations = 0;
            for (int i = 1; i < n; ++i) {
                double dist = stations[i] - stations[i-1];
                neededStations += ceil(dist / maxDist) - 1;
            }
            return neededStations <= k;
        };

        while (right - left > 1e-6) {
            double mid = (left + right) / 2.0;
            if (canPlaceStations(mid)) {
                right = mid;
            } else {
                left = mid;
            }
        }

        return round(right * 100) / 100.0;
    }
};

File: Minimize Max Distance to Gas Station.py
import math
from typing import List

class Solution:
    def numberOfGasStationsRequired(self, dist: float, arr: List[int]) -> int:
        cnt = 0
        for i in range(1, len(arr)):
            gap = arr[i] - arr[i - 1]
            cnt += math.ceil(gap / dist) - 1
        return cnt

    def findSmallestMaxDist(self, stations: List[int], K: int) -> float:
        low, high = 0, max(stations[i + 1] - stations[i] for i in range(len(stations) - 1))

        diff = 1e-6
        while high - low > diff:
            mid = (low + high) / 2.0
            cnt = self.numberOfGasStationsRequired(mid, stations)
            if cnt > K:
                low = mid
            else:
                high = mid

        return high
